src/common/menu.rs

```diff
@@ -70,27 +75,33 @@ pub unsafe fn write_menu() {
 
 const MENU_CONF_PATH: &str = "sd:/TrainingModpack/training_modpack_menu.conf";
 
-pub fn set_menu_from_url(last_url: &str) {
-    unsafe {
-        MENU = get_menu_from_url(MENU, last_url, false);
-        DEFAULTS_MENU = get_menu_from_url(MENU, last_url, true);
-
-        if MENU.quick_menu == OnOff::Off {
-            if is_emulator() {
-                skyline::error::show_error(
-                    0x69,
-                    "Cannot use web menu on emulator.\n\0",
-                    "Only the quick menu is runnable via emulator currently.\n\0",
-                );
-                MENU.quick_menu = OnOff::On;
-            }
+pub unsafe fn set_menu_from_json(message: &str) {
+    if MENU.quick_menu == OnOff::Off {
+        if is_emulator() {
+            skyline::error::show_error(
+                0x69,
+                "Cannot use web menu on emulator.\n\0",
+                "Only the quick menu is runnable via emulator currently.\n\0",
+            );
+            MENU.quick_menu = OnOff::On;
         }
     }
-
-    std::fs::write(MENU_CONF_PATH, last_url).expect("Failed to write menu conf file");
-    unsafe {
-        EVENT_QUEUE.push(Event::menu_open(last_url.to_string()));
-    }
+    if let Ok(message_json) = serde_json::from_str::<MenuJsonStruct>(message) {
+        MENU = message_json.menu;
+        DEFAULTS_MENU = message_json.defaults_menu;
+        std::fs::write(
+            MENU_CONF_PATH,
+            serde_json::to_string_pretty(&message_json).unwrap(),
+        )
+        .expect("Failed to write menu conf file");
+        // EVENT_QUEUE.push(Event::menu_open(message_json)); // TODO
```

Comment:
For this, you can just pass through the string `message`, data processing side can handle the serialization from there.